File: pyregpoly/tgfsr.py
"""
tgfsr.py — Twisted GFSR (TGFSR) generator.

The state consists of r words of w bits (k = w*r total).
Each step applies a twist matrix A to two state words at prescribed
distances, then shifts and XORs the result into the state register.

Characteristic polynomial: computed via Berlekamp-Massey (inherited).

Implements the Generateur abstract base class.
"""

from __future__ import annotations

import sys

from bitvect import BitVect
from generateur import Generateur


class TGFSR(Generateur):
    """
    Twisted Generalized Feedback Shift Register.

    Attributes
    ----------
    k    : int     — degree = w * r  (inherited)
    w    : int     — word size in bits
    r    : int     — number of w-bit words in the state register
    m    : int     — middle distance parameter
    a    : BitVect — twist matrix A (k bits, only first w bits used)
    L    : int     — output resolution in bits
    """

    def __init__(self, w: int, r: int, m: int, a: BitVect, L: int) -> None:
        k = w * r
        super().__init__(k)
        self.w: int = w
        self.r: int = r
        self.m: int = m
        self.a: BitVect = a.copy()
        self.L = L
        self.step = 1
        self.smax = sys.maxsize

    # -- Generateur interface ---------------------------------------------

    @classmethod
    def name(cls) -> str:
        return "TGFSR"

    def display(self) -> None:
        """DispTGFSR: print w, r, m and the twist matrix a."""
        a_word = self.a._val >> (self.a.n - 32) if self.a.n >= 32 else self.a._val
        print(f" w= {self.w:3d}  r= {self.r:3d}  m= {self.m:3d}  a= {a_word:08x}")

    def initialize_state(self, init_bv: BitVect) -> BitVect:
        """InitTGFSR: copy init into state, return first L bits."""
        self.gen_state = BitVect(self.k, 0)
        self.gen_state.copy_part_from(init_bv, self.k)
        retour = BitVect(self.L, 0)
        retour.copy_part_from(init_bv, self.L)
        return retour

    def __next__(self) -> BitVect:
        """
        TGFSR: one recurrence step.

        Algorithm (matching the C code):
          Temp2 = (state << w*(r-m-1)) masked to w bits   [word at distance m+1]
          Temp  = (state << w*(r-1))   masked to w bits   [last word]
          if bit w-1 of Temp == 0:
              Temp = Temp >> 1  XOR  Temp2
          else:
              Temp = Temp >> 1  XOR  Temp2  XOR  A
          Temp masked to w bits
          state = (state >> w)  XOR  Temp
          return first L bits of state
        """
        w = self.w
        r = self.r
        m = self.m
        k = self.k
        state = self.gen_state

        # Extract word at distance (r-m-1) and last word (r-1)
        temp2 = (state << (w * (r - m - 1))).and_mask(w)
        temp  = (state << (w * (r - 1))).and_mask(w)

        # Check LSB of the w-bit word (bit w-1 in public indexing)
        if temp.get_bit(w - 1) == 0:
            temp >>= 1
            temp = temp2 ^ temp
        else:
            temp >>= 1
            temp = temp2.xor3(temp, self.a)

        # Shift state right by w, XOR with masked temp
        state >>= w
        temp = temp.and_mask(w)
        self.gen_state = state ^ temp

        # Return first L bits
        retour = BitVect(self.L, 0)
        retour.copy_part_from(self.gen_state, self.L)
        return retour

    def copy(self) -> "TGFSR":
        new = TGFSR(self.w, self.r, self.m, self.a, self.L)
        new.gen_state = self.gen_state.copy()
        new.smax = self.smax
        new.step = self.step
        return new

    # -- File I/O ---------------------------------------------------------

    @classmethod
    def CreateListFromFile(cls, filename: str, L: int) -> list["TGFSR"]:
        """
        ReadDataTGFSR: read TGFSR parameters from filename.

        File format
        -----------
        Line 1 : type tag  ("tgfsr")
        Line 2 : w  r
        Line 3 : nbgen
        Lines 4 .. nbgen+3:
            hex_a  m
            hex_a is the twist matrix value (stored in vect[0])
            m is the middle distance parameter

        Returns a list of TGFSR instances.
        """
        with open(filename) as f:
            f.readline()                        # skip type tag
            parts = f.readline().split()
            w = int(parts[0])
            r = int(parts[1])
            if w > 32:
                raise ValueError(f"Error: w must be <= 32 bits, got {w}")
            k = w * r
            nbgen = int(f.readline().split()[0])

            generators = []
            for _ in range(nbgen):
                tokens = f.readline().split()
                a_val = int(tokens[0], 16)
                m_val = int(tokens[1])

                # Build a k-bit BitVect with the hex value in the top 32 bits
                # (matching C's vect[0] storage)
                a_bv = BitVect(k, a_val << (k - 32) if k > 32 else a_val)

                gen_L = min(w, L)
                generators.append(cls(w, r, m_val, a_bv, gen_L))

        return generators

File: c_regpoly/src/tgfsr.c
/*
 * tgfsr.c — Twisted GFSR (TGFSR) generator for REGPOLY.
 *
 * Implements the Twisted Generalized Feedback Shift Register. The state
 * consists of r words of w bits; each step applies a twist matrix A to
 * two consecutive words then XORs in the word at distance m.
 * The characteristic polynomial is computed using the ZEN library.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <limits.h>
#include "tgfsr.h"
#include "polynomials.h"

#define TGFSRNAME "TGFSR"

#define TW   ((paramTGFSR *)(Gen->ParamGen))->w
#define TR   ((paramTGFSR *)(Gen->ParamGen))->r
#define TM   ((paramTGFSR *)(Gen->ParamGen))->m
#define TA   ((paramTGFSR *)(Gen->ParamGen))->a
#define ETAT Gen->GenState

/* Returns the name string for the TGFSR generator type. */
char *TGFSRName(void)
{
  return TGFSRNAME;
}

/* Prints the parameters (w, r, m, a) of the TGFSR generator. */
void DispTGFSR(Generateur *Gen)
{
  printf(" w= %3u  r= %3u  m= %3u  a= %08x\n",
         ((paramTGFSR *)(Gen->ParamGen))->w,
         ((paramTGFSR *)(Gen->ParamGen))->r,
         ((paramTGFSR *)(Gen->ParamGen))->m,
         ((paramTGFSR *)(Gen->ParamGen))->a.vect[0]);
}

/* Initializes all parameters of a TGFSR generator with word size w, register
   count r, middle distance m, twist matrix A, and output resolution L.
   Wires all function pointers for this type. */
void InitParamTGFSR(Generateur *Gen, int w, int r, int m, BitVect *A, int L)
{
  ((paramTGFSR *)(Gen->ParamGen))->w = w;
  ((paramTGFSR *)(Gen->ParamGen))->r = r;
  ((paramTGFSR *)(Gen->ParamGen))->m = m;
  CopyBV(&(((paramTGFSR *)(Gen->ParamGen))->a), A);

  Gen->k    = w * r;
  Gen->L    = L;
  Gen->Step = 1;
  Gen->smax = INT_MAX;
  Gen->DispGen   = DispTGFSR;
  Gen->DispName  = TGFSRName;
  Gen->InitGen   = InitTGFSR;
  Gen->Iteration = TGFSR;
  Gen->CopyGen   = CopyTGFSR;
  Gen->AllocGen  = AllocTGFSR;
  Gen->PolyChar  = CharTGFSR;
  PutBVToZero(&(Gen->GenState));
}

/* Copies all fields of TGFSR generator Gen2 into Gen1,
   including a deep copy of the twist matrix. */
void CopyTGFSR(Generateur *Gen1, Generateur *Gen2)
{
  ((paramTGFSR *)(Gen1->ParamGen))->w = ((paramTGFSR *)(Gen2->ParamGen))->w;
  ((paramTGFSR *)(Gen1->ParamGen))->r = ((paramTGFSR *)(Gen2->ParamGen))->r;
  ((paramTGFSR *)(Gen1->ParamGen))->m = ((paramTGFSR *)(Gen2->ParamGen))->m;
  AllocBV(&(((paramTGFSR *)(Gen1->ParamGen))->a), Gen2->k);
  CopyBV(&(((paramTGFSR *)(Gen1->ParamGen))->a),
         &(((paramTGFSR *)(Gen2->ParamGen))->a));

  Gen1->k         = Gen2->k;
  Gen1->L         = Gen2->L;
  Gen1->smax      = Gen2->smax;
  Gen1->Step      = Gen2->Step;
  Gen1->DispGen   = Gen2->DispGen;
  Gen1->DispName  = Gen2->DispName;
  Gen1->InitGen   = Gen2->InitGen;
  Gen1->Iteration = Gen2->Iteration;
  Gen1->PolyChar  = Gen2->PolyChar;
  Gen1->CopyGen   = Gen2->CopyGen;
  Gen1->AllocGen  = Gen2->AllocGen;
  CopyBV(&(Gen1->GenState), &(Gen2->GenState));
}

/* Allocates and returns a new TGFSR generator with state width k. */
Generateur *AllocTGFSR(int k)
{
  Generateur *G;
  G = (Generateur *) calloc(1, sizeof(Generateur));
  AllocGen(G, k);
  G->ParamGen = (paramTGFSR *) calloc(1, sizeof(paramTGFSR));
  AllocBV(&(((paramTGFSR *)(G->ParamGen))->a), k);
  return G;
}

/* Frees all memory associated with the TGFSR generator G. */
void FreeTGFSR(Generateur *G)
{
  FreeBV(&(((paramTGFSR *)(G->ParamGen))->a));
  free(G->ParamGen);
  FreeGen(G);
  free(G);
}

/* Initializes the TGFSR state from the bit vector init and copies
   the L low-order bits into retour. */
void InitTGFSR(Generateur *Gen, BitVect *init, BitVect *retour)
{
  CopyBV(&ETAT, init);
  CopyBVPart(retour, init, Gen->L);
}

/* Performs one TGFSR recurrence step and writes L output bits to retour. */
void TGFSR(Generateur *Gen, BitVect *RETOUR)
{
  BitVect Temp, Temp2;

  AllocBV(&Temp,  Gen->k);
  AllocBV(&Temp2, Gen->k);

  BVLShift(&Temp2, &ETAT, TW * (TR - TM - 1));
  BVLShift(&Temp,  &ETAT, TW * (TR - 1));
  ANDBVMask(&Temp2, &Temp2, TW);
  ANDBVMask(&Temp,  &Temp,  TW);

  if (ValBitBV(&Temp, TW - 1) == 0) {
    BVRShiftSelf(&Temp, 1);
    XORBV(&Temp, &Temp2, &Temp);
  } else {
    BVRShiftSelf(&Temp, 1);
    XOR2BV(&Temp, &Temp2, &Temp, &TA);
  }

  BVRShiftSelf(&ETAT, TW);
  ANDBVMask(&Temp, &Temp, TW);
  XORBV(&ETAT, &ETAT, &Temp);
  CopyBVPart(RETOUR, &ETAT, Gen->L);
  FreeBV(&Temp);
  FreeBV(&Temp2);
}

/* Reads TGFSR parameters from filename and adds each generator to component E.
   Exits if the file cannot be opened, w > 32, or allocation fails. */
void ReadDataTGFSR(Component *E, char *filename, boolean same, int L)
{
  int i, nbgen, m;
  int w, r;
  BitVect a_BV;
  FILE *f;
  Generateur *p;
  char gentype[25];

  f = fopen(filename, "r");
  if (f == NULL) {
    printf("File %s not found.\n", filename);
    exit(1);
  }
  fscanf(f, "%s", gentype);
  ReadLn(f);
  fscanf(f, "%d", &w);
  fscanf(f, "%d", &r);
  if (w > 32) {
    printf("Error: w must be <= 32 bits.\n");
    exit(1);
  }
  ReadLn(f);
  fscanf(f, "%d", &nbgen);
  AllocGensInComponent(E, nbgen, same);
  AllocBV(&a_BV, w * r);
  if ((p = AllocTGFSR(r * w)) == NULL) {
    printf("Error in ReadDataTGFSR()\n");
    exit(1);
  }
  ReadLn(f);
  for (i = 0; i < nbgen; i++) {
    fscanf(f, "%x", &(a_BV.vect[0]));
    fscanf(f, "%d",  &m);
    InitParamTGFSR(p, w, r, m, &a_BV, intmin(w, L));
    ReadLn(f);
    AddGenInComponent(E, p);
  }
  FreeBV(&a_BV);
  fclose(f);
  FreeTGFSR(p);
}

/* Generates nb random TGFSR parameter sets with word size w and register count r
   that yield primitive characteristic polynomials, and writes them to filename. */
void ProduceParamsTGFSR(char *filename, int nb, int w, int r)
{
  Generateur *p;
  BitVect a, BC, Dummy;
  FILE *f;
  int i, nbcoeff, c, count = 0, *Poly, m;

  SetMRG32k3a((double)time(NULL), (double)time(NULL), (double)time(NULL),
              (double)time(NULL), (double)time(NULL), (double)time(NULL));
  f = fopen(filename, "w");
  Poly = (int *) malloc((w * r + 1) * sizeof(int));
  fprintf(f, "tgfsr\n%d %d\n", w, r);
  fprintf(f, "%d\n", nb);

  while (count < nb) {
    for (i = 0; i <= w; i++)
      Poly[i] = 0;
    Poly[0] = 1;
    Poly[w] = 1;
    nbcoeff = 2;
    for (i = 1; i < w; i++)
      if ((Poly[i] = MRG32k3a() % 2) == 1)
        nbcoeff++;
    if (!(nbcoeff % 2)) {
      c = MRG32k3a() % (w - 1) + 1;
      if (Poly[c])
        Poly[c] = 0;
      else
        Poly[c] = 1;
    }
    if (IrreduciblePolynomial(Poly, w)) {
      AllocBV(&a, w * r);
      PutBVToZero(&a);
      AllocBV(&BC, w * r);
      BVCanonic(&BC, 0);
      for (i = 0; i < w; i++) {
        if (Poly[i])
          XORBV(&a, &a, &BC);
        BVRShiftSelf(&BC, 1);
      }
      FreeBV(&BC);
      if ((p = AllocTGFSR(w * r)) == NULL) {
        printf("Error in ProduceParamsTGFSR()\n");
        exit(1);
      }
      m = (MRG32k3a() % (r - 1)) + 1;
      InitParamTGFSR(p, w, r, m, &a, WL);
      AllocBV(&Dummy, (w * r) + 1);
      CharTGFSR(p, Poly, &Dummy);

      if (PrimitifPolynomial(Poly, w * r)) {
        fprintf(f, "%08x %d\n", a.vect[0], m);
        count++;
      }
      FreeBV(&a);
      FreeBV(&Dummy);
      FreeGen(p);
    }
  }
  fclose(f);
}

/* Computes the characteristic polynomial of the TGFSR using the ZEN library.
   Stores the result in coeff[] and BVPoly. */
void CharTGFSR(Generateur *Gen, int *coeff, BitVect *BVPoly)
{
  int j;
  ZENPoly tntm, res, temp;
  ZENRing F2;
  BigNum two;
  BigNumLength twol;
  ZENElt ONE, Coeff;

  ZBNReadFromString(&two, &twol, "2", 10);
  if (ZENBaseRingAlloc(F2, two, twol)) {
    printf("Error in CharTGFSR(): ZENBaseRingAlloc failed\n");
    exit(1);
  }
  ZENEltAlloc(ONE, F2);
  ZENEltSetToOne(ONE, F2);
  ZENEltAlloc(Coeff, F2);

  ZENPolyAlloc(temp, TR * TW, F2);
  ZENPolySetToZero(temp, F2);
  ZENPolyAlloc(res, TR * TW, F2);

  /* t^n + t^m */
  ZENPolyAlloc(tntm, TR * TW, F2);
  ZENPolySetToXi(tntm, TR, F2);
  ZENPolySetCoeff(tntm, TM, ONE, F2);

  ZENPolySetToZero(res, F2);
  for (j = 0; j < TW; j++)
    if (ValBitBV(&TA, j)) {
      PowPoly(&temp, &tntm, &F2, j);
      ZENPolyAdd(res, temp, F2);
    }
  PowPoly(&temp, &tntm, &F2, TW);
  ZENPolyAdd(res, temp, F2);

  for (j = 0; j <= TR * TW; j++) {
    ZENPolyGetCoeff(Coeff, res, j, F2);
    coeff[j] = ZENEltAreEqual(Coeff, ONE, F2) ? 1 : 0;
  }

  ZENEltFree(ONE, F2);
  ZBNF(two);
  ZENPolyFree(tntm, F2);
  ZENPolyFree(res,  F2);
  ZENPolyFree(temp, F2);
  if (ZENRingClose(F2)) {
    printf("Error in CharTGFSR(): ZENRingClose failed\n");
    exit(1);
  }
  PutBVToZero(BVPoly);
  for (j = 0; j <= Gen->k; j++)
    PutBitBV(BVPoly, j, coeff[j]);
}
